packages/nextjs/pages/index.tsx

```diff
@@ -113,9 +119,10 @@ export const getServerSideProps: GetServerSideProps = async ({ res }) => {
   return {
     props: {
       // urqlState is a keyword here so withUrqlClient can pick it up.
-      urqlState: ssrCache.extractData(),
+      urqlState: ssrCache.extractData() as SSRData<GetProductsAndCategoriesQuery>,
     },
+    [Symbol.for("e2eData")]: query?.data,
   };
-};
+});
```

Comment:
# Open Discussion: `getPageData` versus `getServerSideProps`

To write good E2E tests against **production data**, the test needs to have access to that data.
Currently, I have implemented 2 different ways to access that data, and I'm looking for feedback on both approaches.
# TL;DR;
- `getServerSideProps` extracts the data from the `__NEXT_DATA__.props` field but `urqlState` messes with that
- `getPageData` makes an API call to retrieve the data, but it's more complex and has overhead and hacks

## `getServerSideProps`
In this approach, the E2E test visits the page, then retrieves `window.__NEXT_DATA__.props` to access the server-side props.
- It's normally quite easy to make these props type-safe (by using the return type of `[page].getServerSideProps`)
- However, MOST of the time in this app, the only server-side prop is `urqlState` as you see here.  In order to access the data, we must parse the `urqlState`, and to make the results strongly-typed, we have to do this `as SSRData<...>` cast.

## `getPageData`
The API is nearly identical, but the approach is quite different.  With this approach, we expose an HTTP endpoint, `/api/e2e-tests/page-data`.  This endpoint calls the page's actual `getServerSideProps` method, and returns the results over HTTP.
- This allows the E2E test to access the raw data
- This also allows the `getServerSideProps` call to return _different_ data.  For example, instead of only returning the `urqlState`, we can also return the raw `query.data` ... so we don't have to bother extracting the `urqlState`.  
- One caveat here is that `getServerSideProps` doesn't allow you to return non-standard fields, so I can't just add the `e2eData: query.data` to the return value.  I also don't want to add extra, unused data as a `props`, because it would get serialized.  So, I resorted to the caveat here, wrapping the value behind a `Symbol`, which doesn't throw warnings, but can still be accessed.
- Another caveat: getting the server-side data via separate API call means the data might not always be 100% in-sync


# Thoughts?

Reply:
I decided the best approach was `getServerSideProps`, using `__NEXT_DATA__`.  It's very clean and works everywhere.